lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs

```diff
@@ -100,43 +107,86 @@ public byte[] Call(byte[] memento)
                 iterNo++;
-
-                using (
-                    var message = new MapInputWithControlMessage<TMapInput>(updateResult.MapInput,
-                        MapControlMessage.AnotherRound))
+                try
                 {
-                    _dataAndControlMessageSender.Send(message);
+                    using (
+                        var message = new MapInputWithControlMessage<TMapInput>(updateResult.MapInput,
+                            MapControlMessage.AnotherRound))
+                    {
+                        _dataAndControlMessageSender.Send(message);
+                    }
+
+                    var input = _dataReceiver.Reduce(_cancellationSource);
+
+                    if (_invokeGC)
+                    {
+                        Logger.Log(Level.Verbose, "Calling Garbage Collector");
+                        GC.Collect();
+                        GC.WaitForPendingFinalizers();
+                    }
+
+                    updateResult = _updateTask.Update(input);
+
+                    if (updateResult.HasResult)
+                    {
+                        _resultHandler.HandleResult(updateResult.Result);
+                    }
                 }
-
-                var input = _dataReceiver.Reduce();
-
-                if (_invokeGC)
+                catch (OperationCanceledException e)
                 {
-                    Logger.Log(Level.Verbose, "Calling Garbage Collector");
-                    GC.Collect();
-                    GC.WaitForPendingFinalizers();
+                    Logger.Log(Level.Warning, "Received OperationCanceledException in UpdateTaskHost with message: {0}.", e.Message);
+                    break;
                 }
-
-                updateResult = _updateTask.Update(input);
-
-                if (updateResult.HasResult)
+                catch (IOException e)
                 {
-                    _resultHandler.HandleResult(updateResult.Result);
+                    Logger.Log(Level.Error, "Received IOException in UpdateTaskHost with message: {0}.", e.Message);
+                    if (!_cancellationSource.IsCancellationRequested)
+                    {
+                        throw new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError);
+                    }
+                    break;
+                }
+                catch (TcpClientConnectionException e)
+                {
+                    Logger.Log(Level.Error, "Received TcpClientConnectionException in UpdateTaskHost with message: {0}.", e.Message);
+                    if (!_cancellationSource.IsCancellationRequested)
+                    {
+                        throw new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError);
+                    }
+                    break;
                 }
             }
 
-            if (!_taskCloseCoordinator.ShouldCloseTask())
+            if (!_cancellationSource.IsCancellationRequested)
```

Comment:
Why not put a single try-catch statement on the full Update task except the last statement -  _taskCloseCoordinator.SignalTaskStopped(); ? Wouldn't that avoid this duplication issue?